dips/dip-16.md

```diff
@@ -0,0 +1,120 @@
+---
+dip: 16
+title: Transaction Scripts
+authors: Sam Blackshear (@sblackshear), Todd Nowacki (@tnowacki), Bob Wilson (@bob-wilson)
+status: Draft
+type: Informational
+created: 3/24/2021
+---
+
+Diem Framework transactions contain Move transaction scripts that change on-chain state (e.g., balances) by invoking functions of Move modules published on-chain. This DIP explains both the mechanics of transaction scripts and the process for adding, removing, and changing scripts. This DIP does *not* document the specific scripts that are available to Diem users, but the complete list of scripts can be found in the [Diem developer docs](https://github.com/diem/diem/blob/main/language/diem-framework/transaction_scripts/doc/transaction_script_documentation.md).
+
+### Transaction payloads
+
+A Diem transaction includes a `TransactionPayload` that contains the logic for performing an on-chain update:
+
+```
+pub enum TransactionPayload {
+    /// A system maintenance transaction.
+    WriteSet(WriteSetPayload),
+    /// A transaction that executes code.
+    Script(Script),
+    /// A transaction that publishes code.
+    Module(Module),
+    /// A transaction that executes an existing script function published on-chain.
+    ScriptFunction(ScriptFunction),
+}
+```
+
+The two `TransactionPayload` variants relevant to this DIP are `ScriptFunction` (which contains a pointer to a function already published in an on-chain Move module) and `Script` (a standalone, single-function bytecode program).
+
+## Script signature requirements
+
+Both script functions and the single function in a transaction script bytecode file have the following requirements. For a function` f<ability_params>(param_types): ret_types`:
```

Comment:
I'm confused by this ability_params, because in our rust code, the ability is refer to copy, drop, store and key, is this ability_params same thing or they are generic type info?